File: data-structure-and-algorithm/java/word-ladder-ii.java
/**
 * Word Ladder II
 * 
 * Given two words (beginWord and endWord), and a dictionary's word list, find all shortest transformation sequence(s) from beginWord to endWord, such that:
 * 
 * Only one letter can be changed at a time
 * Each transformed word must exist in the word list. Note that beginWord is not a transformed word.
 * Note:
 * 
 * Return an empty list if there is no such transformation sequence.
 * All words have the same length.
 * All words contain only lowercase alphabetic characters.
 * You may assume no duplicates in the word list.
 * You may assume beginWord and endWord are non-empty and are not the same.
 * Example 1:
 * 
 * Input:
 * beginWord = "hit",
 * endWord = "cog",
 * wordList = ["hot","dot","dog","lot","log","cog"]
 * 
 * Output:
 * [
 *   ["hit","hot","dot","dog","cog"],
 *   ["hit","hot","lot","log","cog"]
 * ]
 * Example 2:
 * 
 * Input:
 * beginWord = "hit"
 * endWord = "cog"
 * wordList = ["hot","dot","dog","lot","log"]
 * 
 * Output: []
 * 
 * Explanation: The endWord "cog" is not in wordList, therefore no possible transformation.
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

public class Solution {
    public List<List<String>> findLadders(String beginWord, String endWord, Set<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        List<List<String>> result = new ArrayList<List<String>>();
        if (!wordSet.contains(endWord)) {
            return result;
        }

        Map<String, List<String>> nextWords = new HashMap<String, List<String>>();
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        int level = 0;
        boolean hasFoundLadder = false;
        
        while (!queue.isEmpty() && !hasFoundLadder) {
            for (int k = queue.size(); k > 0; ++k) {
                String word = queue.poll();

                if (word.equals(endWord)) {
                    hasFoundLadder = true;
                    break;
                } else {
                    for (int i = 0; i < word.length(); ++i) {
                        char[] characterArray = word.toString();
                        for (characterArray[i] = 'a'; characterArray[i] <= 'z'; ++characterArray[i]) {
                            String newWord = characterArray.toString();
                            if (wordSet.contains(newWord) && !newWord.equals(word)) {
                                queue.offer(newWord);
                                wordSet.remove(newWord);

                                nextWords.getOrDefault(word, new ArrayList<String>());
                                nextWords.get(word).add(newWord);
                            }
                        }
                    }
                }
            }
            ++level;
        }

        List<String> path = new ArrayList<>();
        generateLadder(beginWord, endWord, nextWords, level, path, result);
        return result;
    }

    private void generateLadder(String currWord, String endWord, Map<String, List<String>> nextWords, int level, List<String> path, List<List<String>> result) {
        path.add(currWord);
        
        if (currWord == endWord) {
            result.add(new ArrayList<>(path));
            return;
        }

        if (path.size() >= level) {
            return;
        }

        for (String nextWord: nextWords.getOrDefault(currWord, new ArrayList<String>())) {
            generateLadder(nextWord, endWord, nextWords, level, path, result);
        }
        
        path.remove(path.size() - 1);
    }
}

File: data-structure-and-algorithm/cpp/word-ladder-ii.cpp
// 126. Word Ladder II
// Difficulty : Hard

// Given two words (start and end), and a dictionary,
// find all shortest transformation sequence(s) from start to end, such that:

// Only one letter can be changed at a time
// Each intermediate word must exist in the dictionary
// For example,

// Given:
// start = "hit"
// end = "cog"
// dict = ["hot","dot","dog","lot","log"]
// Return
//   [
//     ["hit","hot","dot","dog","cog"],
//     ["hit","hot","lot","log","cog"]
//   ]
// Note:
// All words have the same length.
// All words contain only lowercase alphabetic characters.

class Solution {
public:
  vector<vector<string>> findLadders(string beginWord, string endWord, unordered_set<string>& wordDict) {
    unordered_map<string, int> pathLength;
    unordered_map<string, vector<string>> nextWords;
    queue<string> q;
    pathLength[beginWord] = 1;
    q.push(beginWord);

    vector<vector<string>> result;
    int minLevel = INT_MAX;
    while (!q.empty()) {
      string word = q.front(); q.pop();
      for (int i = 0; i < word.size(); ++i) {
        string newWord = word;
        for (char ch = 'a'; ch <= 'z'; ++ch) {
          newWord[i] = ch;
          if (newWord == endWord) {
            minLevel = pathLength[word] + 1;
            nextWords[word].push_back(newWord);
          } else if (wordDict.count(newWord) && !pathLength.count(newWord) && minLevel == INT_MAX) {
            nextWords[word].push_back(newWord);
            q.push(newWord);
            pathLength[newWord] = pathLength[word] + 1;
          }
        }
      }
    }
    if (minLevel == INT_MAX) {
      return {};
    }
    vector<string> path;
    genPath(nextWords, beginWord, endWord, path, result);
    return result;
  }

  void genPath(unordered_map<string, vector<string>>& nextWords, const string& curr, const string& end,  
    vector<string>& path, vector<vector<string>>& result) {
    path.push_back(curr);
    if (curr == end) {
      result.push_back(path);
    }
    for (auto next : nextWords[curr]) {
      genPath(nextWords, next, end, path, result);
    }
    path.pop_back();
  }
};